components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java

```diff
@@ -98,6 +106,140 @@ public UserSession getSession(String sessionId) throws SessionManagementServerEx
+                sqlQuery = MessageFormat.format(SQLQueries.SQL_LOAD_SESSIONS_H2, sqlFilters[0],
+                        appIdFilter, sqlFilters[2], sqlFilters[3], sqlOrder, limit);
+            } else if (JdbcUtils.isMySQLDB(JdbcUtils.Database.SESSION)) {
+                sqlQuery = MessageFormat.format(SQLQueries.SQL_LOAD_SESSIONS_MYSQL, sqlFilters[0],
+                        appIdFilter, sqlFilters[2], sqlFilters[3], sqlOrder, limit);
+            } else if (JdbcUtils.isOracleDB(JdbcUtils.Database.SESSION)) {
+                sqlQuery = MessageFormat.format(SQLQueries.SQL_LOAD_SESSIONS_ORACLE, sqlFilters[0],
+                        appIdFilter, sqlFilters[2], sqlFilters[3], sqlOrder, limit);
+            } else if (JdbcUtils.isMSSqlDB(JdbcUtils.Database.SESSION)) {
+                sqlQuery = MessageFormat.format(SQLQueries.SQL_LOAD_SESSIONS_MSSQL, sqlFilters[0],
+                        appIdFilter, sqlFilters[2], sqlFilters[3], sqlOrder, limit);
+            } else if (JdbcUtils.isPostgreSQLDB(JdbcUtils.Database.SESSION)) {
+                sqlQuery = MessageFormat.format(SQLQueries.SQL_LOAD_SESSIONS_POSTGRESQL, sqlFilters[0],
+                        appIdFilter, sqlFilters[2], sqlFilters[3], sqlOrder, limit);
+            } else if (JdbcUtils.isDB2DB(JdbcUtils.Database.SESSION)) {
+                sqlQuery = MessageFormat.format(SQLQueries.SQL_LOAD_SESSIONS_DB2, sqlFilters[0],
+                        appIdFilter, sqlFilters[2], sqlFilters[3], sqlOrder, limit);
+            } else {
+                throw new UserSessionException("Error while loading sessions from DB: Database driver could not " +
+                        "be identified or not supported.");
+            }
+        } catch (DataAccessException e) {
+            throw new UserSessionException("Error while loading sessions from DB: Database driver could not be " +
+                    "identified or not supported.", e);
+        }
+
+        JdbcTemplate jdbcTemplate = JdbcUtils.getNewTemplate(JdbcUtils.Database.SESSION);
+        try {
+            Map<String, Application> finalAppDetails = appDetails;
+            userSessionsList = jdbcTemplate.executeQuery(
+                    sqlQuery,
+                    ((resultSet, rowNumber) -> SessionMgtUtils.parseSessionSearchResult(resultSet, finalAppDetails)),
+                    preparedStatement -> {
+                        preparedStatement.setLong(1, FrameworkUtils.getCurrentStandardNano());
+                        preparedStatement.setInt(2, tenantId);
+                    });
+
+            // Application details will be incomplete if an application filter is not provided. In that case
+            // requires to query for missing application details.
+            if (!userSessionsList.isEmpty() && finalAppDetails.isEmpty()) {
+                for (UserSession userSession : userSessionsList) {
+                    generateApplicationFromAppID(userSession.getApplications());
```

Comment:
Check whether we can do this in a less no of DB query. 
For examples when we are having 20 sessions, but only 3 applications, Can do it in 3 SP_APP queries instead of 20